### PR TITLE
fix: harden data sync, API limits, startup readiness, renderer, and web assets

### PR DESCRIPTION
**Pull request by Rhiz3K**
## Summary

Review-driven hardening release (1.2.40). Each topic is a separate commit for review; the PR is meant to be squash-merged.

- **Season sync**: an unpublished next season no longer fails `scripts/update_seasons.py`, and the workflow opens a PR for updated seasons before reporting a failed year. The weekly sync has failed since 2026-07-27, so the bundled 2026 calendar is stale (Sepang on 2026-10-04 is missing). Run the "Update F1 Season Data" workflow manually after merging.
- **API**: rate limit before `ADMIN_API_TOKEN` validation, compare tokens as bytes (non-ASCII header returned HTTP 500), validate perf-metric `page_path`, collapse unknown pages into `/other`, lower the default beacon quota to 30/min.
- **Startup**: generate calendars before the historical catch-up and the all-circuit weather fetch so `/health/ready` does not sit in `starting` past the container health-check window.
- **Upstream fan-out**: cache raw Jolpica season/round payloads (positive + negative TTL, per-key coalescing), keep degraded team rosters for five minutes, parse static season files once per on-disk version, log per-request lookups at debug.
- **Changelog page**: five-minute negative cache for failed GitHub version lookups.
- **Renderer**: resample 1-bit track maps in grayscale (Pillow silently uses NEAREST for mode `1`), C-level row activity counts and nibble packing. Cold warmup 9.4 s → 0.86 s, 4-bit BMP encode 102 ms → 12 ms.
- **reset-db**: strict shell mode, SQLite busy timeout, refuse `all` while WAL/SHM sidecars exist unless `--force`, docs for running against a stopped service.
- **Web**: escape translations injected via `innerHTML`, render-blocking versioned stylesheets (no more unstyled first paint / CLS), `aria-label` on the icon-only close button, `for=` on sidebar controls, escaped Umami website id.
- **CI / container**: Dependabot covers composite actions, aligned `setup-uv` pin, removed unused `id-token` permission, `exec uvicorn`, `cap_drop: ALL` + `no-new-privileges`, Sentry release tag, contributor docs use the uv dependency group.

## Testing

- `uv run --locked ruff check .`, `ruff format --check`, `mypy`, `interrogate` (100 %)
- `uv run --locked pytest tests/ -m "not benchmark" --cov=app --cov-branch` → 1141 passed, 99.98 % coverage
- `uv run --locked diff-cover coverage.xml --compare-branch=main --fail-under=100` → 100 %
- `uv run --locked python -m app.utils.release_validation` → OK (1.2.40 > v1.2.39)
- `scripts/update_seasons.py --years 2027` against the live API exits 0 ("not published yet"); `--years 2026` produced 25 races including Sepang (asset intentionally not committed here)
- `scripts/reset_db.sh` exercised through `tests/test_reset_db_script.py`

## Screenshots

No rendered-layout changes; 1-bit track maps only lose the NEAREST dashing when the preprocessed asset is taller than the runtime box.

## Checklist

- [x] I read the Code of Conduct and this change follows it.
- [x] I added or updated documentation as needed.
- [x] I verified error handling and logging are appropriate for FastAPI endpoints.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Security**
  * Strengthened protection for analytics, performance metrics, and containerized deployments.
  * Added validation and normalization for submitted page paths.
  * Reduced the unauthenticated performance-metrics request limit.

* **Performance & Reliability**
  * Improved F1 data loading, failure handling, track-map rendering, asset freshness, and weather generation.
  * Added monitoring for failed season-data updates and track-asset coverage.

* **Accessibility**
  * Improved form-label associations and mobile settings controls with translated labels.

* **New Features**
  * Added validated tooling for importing official track artwork and generating palette variants.

* **Maintenance**
  * Added Sepang calendar data and corrected Miami practice timing.
  * Improved database reset handling and updated development documentation.

### CODE REVIEW

File: app/services/f1_service.py [hunk 1]
```diff
@@ -317,72 +376,89 @@ async def get_season_races(self, year: int) -> list[dict]:
...
-            round_num: The round number
+            year: The season year (e.g., 2025)
 
         Returns:
-            Dictionary with race data including converted times, or None if failed
+            List of race dictionaries with basic info
         """
+        races = await self._get_remote_season_payload(year)
+        if races is None:
+            return []
```
[the remark this answers is not in the document]
Comment (Rhiz3K) [reply app/services/f1_service.py:448]: Addressed in 885b6c2. Empty upstream calendars now remain an empty list, so get_season_races() continues to _merge_static_cancelled_races; test_get_season_races_merges_static_cancellations_for_empty_upstream_calendar covers both the initial and cached calls. None is reserved for an actual fetch failure, intentionally matching the pre-extraction behavior where the outer exception handler returned [] before the merge. The production API is also static-first.

File: app/services/f1_service.py [hunk 2]
```diff
@@ -40,6 +44,85 @@
...
+    def lookup(self, key: tuple) -> tuple[bool, object | None]:
+        """Return ``(hit, payload)``; a remembered empty answer or failure hits too."""
+        if key in self._positive:
+            return True, self._positive[key]
+        if key in self._negative:
+            return True, self._negative[key]
+        return False, None
+
+    def store(self, key: tuple, payload: object | None, *, valid: bool) -> None:
+        """Cache valid data positively, empty answers negatively, and ignore malformed data."""
```
Comment (Rhiz3K) [inline app/services/f1_service.py]: **[suggestion]** `_RemotePayloadCache.store` caches valid payloads for an hour and empty/failed payloads for 60s, but deliberately ignores malformed non-empty payloads. `_is_valid_remote_season_payload` / `_is_valid_remote_race_payload` fail the whole payload if any row cannot `Race.model_validate` or parse a datetime, so a single bad Jolpica row is never stored. Coalescing only helps concurrent waiters on the same lock; sequential requests after the lock releases all miss and enqueue another paced upstream fetch — the fan-out this cache was meant to stop, just on the malformed path rather than the empty/error path.

**Suggestion:** Negative-cache malformed payloads for `REMOTE_NEGATIVE_CACHE_TTL_SECONDS` as well (keep them out of the positive cache). Callers already skip bad rows and merge static cancellations, so a brief “do not refetch” window would bound bursts without pinning a bad calendar for an hour.

File: app/routes/pages.py [hunk 3]
```diff
@@ -506,7 +510,7 @@ async def _changelog_handler(request: Request, ui_lang: str) -> HTMLResponse:
     changelog_html = await asyncio.to_thread(_load_changelog_html, CHANGELOG_PATH)
 
     version_info = get_cached_version()
-    if version_info is None:
+    if version_info is None and not version_fetch_recently_failed():
```
Comment (Rhiz3K) [inline app/routes/pages.py:513]: **[suggestion]** The changelog negative cache only trips when `fetch_version_info` returns an all-`None` `VersionInfo` and sets `_version_fetch_failed_at`. `refresh_version_info` still swallows unexpected exceptions and returns `None` without opening that window, and `_changelog_handler` treats that `None` the same as “no cache yet”. The common GitHub 503 path is covered; a failure in `get_shared_http_client` (or anything else raised before the all-`None` branch) still repeats two GitHub calls on every `/changelog` view for the outage.

**Suggestion:** Set `_version_fetch_failed_at` in `refresh_version_info`’s `except` (and/or whenever it returns `None` with an empty cache), so the handler’s `version_fetch_recently_failed()` check covers every uncached failure mode.

File: app/templates/configure.html [hunk 4]
```diff
@@ -71,14 +71,17 @@
   >
     <!-- Close button (mobile only) -->
     <button
+      type="button"
       onclick="toggleSidebar()"
+      aria-label="{{ configure_ui_text.settingsBtn }}"
```
Comment (Rhiz3K) [inline app/templates/configure.html]: **[suggestion]** The icon-only sidebar close control is labeled with `configure_ui_text.settingsBtn` (“Settings” / “Nastavení” / …), which is the same string as the button that *opens* the sidebar. The control is an X that calls `toggleSidebar()`, and `base.html` already names the analogous mobile-nav control `"Close menu"`. Screen readers will announce this as Settings rather than Close. Language switches also never update this server-rendered `aria-label`.

**Suggestion:** Use a close-specific string (hardcoded `"Close"` like the nav menu, or a new `configure_ui_text` key) and keep it in sync when the UI language changes.

**Review by Rhiz3K — COMMENTED** [review]
## Summary

This is a careful, well-tested hardening release: unpublished next-season sync, Jolpica payload coalescing, `/health/ready` generation-before-catch-up, operational API token/rate-limit ordering, page_path allowlisting, renderer/BMP speedups, reset-db guards, and web XSS/a11y/CSS cache-busting all hang together and match the surrounding call sites. I did not find a correctness or security defect in the auth compare, XSS escaping, WAL reset guard, or startup-order change. The remaining risks are small residual holes (malformed-payload fan-out, changelog negative cache on unexpected `None`) and one mislabeled close control.

## Issue counts by severity

- bugs: 0
- suggestions: 3
- nits: 0

File: app/services/f1_service.py (hunk 2 above)
Comment (Rhiz3K) [reply app/services/f1_service.py]: Fixed in 1048276. Truthy malformed season and round payloads now go into the 60-second negative cache instead of remaining uncached, while only validated payloads enter the one-hour positive cache. The season and round regression tests now assert that repeated calls perform one upstream fetch and that a later fetch can recover once the short-lived entry is gone.

File: app/routes/pages.py (hunk 3 above)
Comment (Rhiz3K) [reply app/routes/pages.py:513]: Fixed in 44c729f. refresh_version_info() now records _version_fetch_failed_at for unexpected exceptions before returning None, so the changelog handler's existing negative-cache guard covers this failure path as well. The regression test pins the clock and verifies that the failure window opens.

File: app/templates/configure.html (hunk 4 above)
Comment (Rhiz3K) [reply app/templates/configure.html]: Fixed in d8b9c1e. The close control now uses a dedicated closeSettings label, updates its aria-label through setUiLanguage(), and ships localized text for all 13 supported locales. Rendering tests cover both the English and Czech accessible names, and the translation key-parity test passes.

**Review by Rhiz3K — COMMENTED** [review]
## Code Review: Follow-up Hardening & Edge-Case Findings

Overall, PR #124 is in excellent shape (100% diff-coverage, all 1,152 tests passing, types and docstrings clean). During an in-depth review of the diff and upstream data contracts, I identified one edge-case bug and a few small suggestions:

---

### 1. 🐛 [Edge Case] Handle `"time": null` in Jolpica season payloads

In [`app/services/f1_service.py`](file:///app/services/f1_service.py):
```python
# line 110 (_is_valid_remote_season_payload)
race_time = race.get("time", DEFAULT_SESSION_TIME_UTC)

# line 455 (get_season_races)
race_time_str = race.get("time", DEFAULT_SESSION_TIME_UTC)
```

**Issue:**  
For upcoming or provisional races where session start times have not yet been announced, Jolpica/Ergast returns `"time": null` in the JSON payload (parsed as `None` in Python).  
Because the `"time"` key exists, `race.get("time", DEFAULT_SESSION_TIME_UTC)` returns `None` instead of falling back to the default.

Consequently:
1. `f"{race_date}T{race_time}"` evaluates to `"{race_date}TNone"`.
2. `datetime.fromisoformat(...)` raises `ValueError: Invalid isoformat string: '...TNone'`.
3. In `_is_valid_remote_season_payload`, this causes the **entire valid season payload to be rejected as malformed** and placed into the 60-second negative cache.
4. In `get_season_races`, the race is skipped with `Skipping malformed race: ...`, even though the `Race` model explicitly defines `time: Optional[str] = None` and `DEFAULT_SESSION_TIME_UTC` is meant as the fallback.

**Recommended fix:**  
Use `or` (consistent with `_convert_race_times` line 311):
```python
race_time = race.get("time") or DEFAULT_SESSION_TIME_UTC
# and
race_time_str = race.get("time") or DEFAULT_SESSION_TIME_UTC
```

---

### 2. 💡 [Syntax / Style] Parenthesize multiple exception types in `except` clauses

In [`app/services/f1_service.py:112`](file:///app/services/f1_service.py#L112) and [`scripts/update_seasons.py:45`](file:///scripts/update_seasons.py#L45):
```python
except AttributeError, TypeError, ValueError, ValidationError:
```
While Python 3.14's parser accepts comma-separated exceptions without enclosing parentheses, PEP 8 and standard Python convention across versions require a tuple:
```python
except (AttributeError, TypeError, ValueError, ValidationError):
```
Parenthesizing ensures consistency with line 488 of the same file.

---

### 3. 💡 [CLI / Tooling] UX improvements for `scripts/reset_db.sh`

The WAL/SHM sidecar check and SQLite busy timeout are great additions. For automated or scripting use cases:
- Add `-h` / `--help` in the argument loop to print `usage` and exit with code `0` (currently unhandled args exit with code `1`).
- Add a `-y` / `--yes` flag to bypass the interactive `confirm()` prompt when invoked from automated maintenance scripts.

---

### 4. 📝 [Note] Fresh container start without DB cache

In [`app/services/scheduler.py`](file:///app/services/scheduler.py), moving image generation before the historical refresh and circuit weather fetch solves the container health-check timeout nicely.  
On a brand new installation with an empty SQLite DB, the initial generated image will not include weather until the hourly scheduler runs. Consider triggering a quiet background regeneration once `_fetch_all_circuits_weather()` finishes if the weather database was initially empty.

**Comment by Rhiz3K** [comment]
Addressing all four items from the Gemini follow-up review:

1. **`"time": null` — fixed in `2b014f8`.** Both the season-payload validator and list conversion now use `race.get("time") or DEFAULT_SESSION_TIME_UTC`. A regression test verifies that an explicit null remains a valid payload and produces `12:00:00Z`.
2. **Unparenthesized exception lists — resolved without a code change.** This project requires and targets Python 3.14. PEP 758 explicitly supports `except A, B:`, and Ruff confirms both affected files are correctly formatted. The parenthesized form later in the service is required because that clause also uses `as e`.
3. **`reset-db` help / non-interactive mode — deferred, no change in this PR.** This is an optional CLI enhancement rather than a correctness issue. The command is intentionally interactive for destructive operations, while `--force` only bypasses the active-database guard. A `--yes` mode would expand the destructive automation surface and should be considered separately with explicit documentation and tests.
4. **Post-weather regeneration — resolved without a code change.** Initial generation does not depend only on the SQLite cache: `_collect_and_generate()` calls `get_weather_context()`, whose cache-miss path fetches current and race weather directly. The later all-circuit fetch populates the broader cache, and the registered hourly generation handles subsequent refreshes. An empty DB alone therefore does not produce the stated weatherless-start condition.

Targeted verification: **33 F1 service tests passed**, plus Ruff, Ruff formatting, mypy, and 100% interrogate coverage.

This review was submitted as one summary rather than separate inline threads, so there are no per-item conversations to resolve; all four items are addressed by this response.
